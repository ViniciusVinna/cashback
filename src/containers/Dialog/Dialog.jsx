import React, { Fragment, useRef, useCallback, useEffect } from 'react';
import PropTypes from 'prop-types';
import { disableBodyScroll, enableBodyScroll } from 'body-scroll-lock';
import { useDispatch, useSelector } from 'react-redux';

import { toggleDialog } from 'actions';

import Portal from 'components/Portal';
import Waves from 'components/Waves';

import { DialogStyled } from './Dialog.styles';

const {
  DialogClose,
  DialogContent,
  DialogHeader,
  DialogOverlay,
  DialogTitle,
  DialogWrapper,
} = DialogStyled;

const Dialog = ({
  children,
  renderComponent,
  title,
}) => {
  const { isDialogOpen, isScrollLocked } = useSelector(state => state.app);
  const dispatch = useDispatch();
  const dialogRef = useRef(null);

  const handleClose = useCallback(() => {
    dispatch(toggleDialog(false));
  }, [dispatch]);

  const handleKeyUp = useCallback((e) => {
    /* istanbul ignore else */
    if (e.keyCode === 27) {
      handleClose();
    }
  }, [handleClose]);

  useEffect(() => {
    /* istanbul ignore else */
    if (isDialogOpen) {
      document.addEventListener('keyup', handleKeyUp);
    }

    return () => {
      document.removeEventListener('keyup', handleKeyUp);
    };
  }, [handleKeyUp, isDialogOpen]);

  useEffect(() => {
    if (isScrollLocked) {
      disableBodyScroll(dialogRef);
    }

    return () => {
      enableBodyScroll(dialogRef);
    };
  }, [isScrollLocked]);

  return (
    <Fragment>
      {isDialogOpen && (
        <Portal id="dialog">
          <DialogWrapper
            data-testid="dialog"
            ref={dialogRef}
            role="dialog"
          >
            <DialogHeader>
              <DialogTitle>{title}</DialogTitle>

              <DialogClose data-testid="dialog-close" onClick={handleClose} />
            </DialogHeader>
            <Waves color="#3388FF" />

            <DialogContent>
              {children}
            </DialogContent>
          </DialogWrapper>
        </Portal>
      )}

      {isDialogOpen && (<DialogOverlay onClick={handleClose} />)}

      {renderComponent && renderComponent}
    </Fragment>
  );
};

Dialog.propTypes = {
  children: PropTypes.node.isRequired,
  renderComponent: PropTypes.node.isRequired,
  title: PropTypes.string.isRequired,
};

export default Dialog;
