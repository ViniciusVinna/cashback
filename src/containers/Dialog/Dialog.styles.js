import styled from '@emotion/styled';
import { css } from '@emotion/core';

import { FiX } from 'react-icons/fi';

export const DialogStyled = {
  DialogContent: styled('div')(({ theme }) => css`
    align-items: center;
    display: flex;
    flex-wrap: wrap;
    padding: ${theme.space.m} ${theme.space.m};
  `),
  DialogClose: styled(FiX)(({ theme }) => css`
    color: ${theme.color.n100};
    cursor: pointer;
    font-size: ${theme.iconSize.m};
  `),
  DialogHeader: styled('div')(({ theme }) => css`
    align-items: center;
    background-color: ${theme.color.b500};
    display: flex;
    flex-wrap: wrap;
    justify-content: space-between;
    margin-bottom: -1px;
    padding: ${theme.space.xs} ${theme.space.m};
  `),
  DialogOverlay: styled('div')(({ theme }) => css`
    backdrop-filter: blur(2px);
    backface-visibility: hidden;
    background-color: ${theme.color.shadow};
    bottom: 0;
    height: 100%;
    left: 0;
    overflow: hidden;
    position: fixed;
    right: 0;
    top: 0;
    width: 100%;
    z-index: calc(${theme.zIndex.overlay});
  `),
  DialogTitle: styled('h4')(({ theme }) => css`
    color: ${theme.color.n100};
    font-size: ${theme.textSize.m};
    font-weight: ${theme.fontWeight.s};
    margin: 0;
  `),
  DialogWrapper: styled('div')(({ theme }) => css`
      backface-visibility: hidden;
      background-color: ${theme.color.n200};
      border-radius: ${theme.borderRadius.l};
      box-shadow: ${theme.boxShadow.modal};
      display: flex;
      flex-direction: column;
      height: auto;
      justify-content: space-between;
      left: 50%;
      overflow-x: hidden;
      overflow-y: auto;
      position: fixed;
      top: 50%;
      transform: translate(-50%, -50%);
      width: calc(100% - ${theme.space.xxxl});
      z-index: ${theme.zIndex.modal};
    `),
};
