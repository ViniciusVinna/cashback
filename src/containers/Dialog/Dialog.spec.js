import React from 'react';
import { useSelector } from 'react-redux';

import Dialog from './Dialog';

let mockStore = {
  app: {
    screenSize: '',
    isDialogOpen: false,
    isScrollLocked: false,
  },
  purchases: {},
  user: {},
};

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useSelector: jest.fn(),
}));

const props = {
  renderComponent: <div>Trigger Component</div>,
  title: 'Something title',
};

describe('Dialog', () => {
  beforeEach(() => {
    useSelector.mockImplementation(callback => callback(mockStore));
  });

  afterEach(() => {
    useSelector.mockClear();
  });

  describe('Dialog', () => {
    it('should render just the trigger component of dialog ', () => {
      const { container } = render(<Dialog {...props}>Child</Dialog>);
      expect(container).toMatchSnapshot();
    });
  });

  describe('Dialog', () => {
    it('should render the dialog and its content ', () => {
      mockStore = {
        ...mockStore,
        app: {
          isDialogOpen: true,
          isScrollLocked: true,
        },
      };

      const { getByText } = setupRender(<Dialog {...props}>Child</Dialog>, 'dialog');

      expect(getByText('Something title')).toBeDefined();
      expect(getByText('Child')).toBeDefined();
    });

    it('should end the dialog when click on the close button', () => {
      const { container, getByTestId } = setupRender(<Dialog {...props}>Child</Dialog>, 'dialog');

      fireEvent.click(getByTestId('dialog-close'));

      expect(container).toMatchSnapshot();
    });

    it('should end the dialog when esc key is pressed', () => {
      const { container } = setupRender(<Dialog {...props}>Child</Dialog>, 'dialog');

      fireEvent.keyUp(container, {
        key: 'Escape',
        code: 'Escape',
        keyCode: 27,
        charCode: 27,
      });

      expect(container).toMatchSnapshot();
    });
  });
});
