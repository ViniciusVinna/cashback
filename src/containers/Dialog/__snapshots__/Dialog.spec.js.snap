// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dialog interactions should end the dialog when click on the close button 1`] = `
<div>
  <div
    class="css-1l3chg2"
  />
  <div>
    Trigger Component
  </div>
</div>
`;

exports[`Dialog interactions should end the dialog when esc key is pressed 1`] = `
<div>
  <div
    class="css-1l3chg2"
  />
  <div>
    Trigger Component
  </div>
</div>
`;

exports[`Dialog no interactions should render properly 1`] = `
<div>
  <div
    class="css-1l3chg2"
  />
  <div>
    Trigger Component
  </div>
</div>
`;
